Bluetooth: Mesh: Friend enc failure

trond-snekvik opened the pull request:
Re-encrypts single-segment application messages when the network seqnum has changed, to avoid encrypting messages with different seqnums in network and transport. This operation is only required for unsegmented messages, as segmented messages don't need to use the same seqnum in network.

Fixes #19265.

trond-snekvik committed:
Bluetooth: Mesh: Expose header parsing outside net

Provides utility function for parsing network headers outside of the
network layer. The primary intended use-case is friendship.

Signed-off-by: Trond Einar Snekvik <Trond.Einar.Snekvik@nordicsemi.no>

trond-snekvik committed:
Bluetooth: Mesh: Expose app key get

Provides a utility function for getting an application key given a
subnet and an app ID. Primary use-case in friendship re-encryption.

Signed-off-by: Trond Einar Snekvik <Trond.Einar.Snekvik@nordicsemi.no>

trond-snekvik committed:
Bluetooth: Mesh: Ensure seqnum match in app/net

Re-encrypts single-segment application messages when the network seqnum
has changed, to avoid encrypting messages with different seqnums in
network and transport. This operation is only required for unsegmented
messages, as segmented messages don't need to use the same seqnum in
network.

Reinstates the special adv data for friend messages to store the app key
index.

Fixes #19265.

Signed-off-by: Trond Einar Snekvik <Trond.Einar.Snekvik@nordicsemi.no>

trond-snekvik committed:
Bluetooth: Mesh: Skip local messages in rx

Local messages are already enqueued for the LPN in the tx path, and
don't have to be added again in the rx path.

Signed-off-by: Trond Einar Snekvik <Trond.Einar.Snekvik@nordicsemi.no>

trond-snekvik commented:
@menglingao1997 FYI

jhedberg commented on subsys/bluetooth/mesh/friend.c:
```diff
@@ -142,7 +148,7 @@ static void friend_clear(struct bt_mesh_friend *frnd)
 		/* Cancel the sending if necessary */
 		if (frnd->pending_buf) {
-			BT_MESH_ADV(frnd->last)->busy = 0U;
+			FRIEND_ADV(frnd->last)->adv.busy = 0U;
```
I don't think this change is needed since `struct b t_mesh_adv adv` is the first member of `struct friend_adv`.

trond-snekvik replied:
That's true, but wouldn't you agree this is more "correct"? As this patch reintroduces the FRIEND_ADV, I don't think this change is out of scope.

jhedberg replied:
> but wouldn't you agree this is more "correct"

Not necessarily. The way the advertising buffer meta-data is designed is that all extensions of it have the generic part first. That's what `adv.c` internally also assumes (it's oblivious to any extensions). So if you'd risk breaking something with this then `adv.c` is also broken.

trond-snekvik replied:
I see, I'll revert.

jhedberg commented on subsys/bluetooth/mesh/friend.c:
```diff
@@ -316,13 +322,124 @@ static struct net_buf *create_friend_pdu(struct bt_mesh_friend *frnd,
+		.data = &buf->data[10],
+		.__buf = &buf->data[10],
+		.size = buf->len - 10,
+	};
```
That's some pretty wild `net_buf_simple` construction 😄 

I wonder if it'd make sense to add a generic `net_buf_simple` helper that initialises a net_buf_simple to point to an existing buffer?

trond-snekvik replied:
Haha, yeah, this felt pretty dirty, but I couldn't find another way to cleanly pull out the sdu without altering the buf. Storing the state and manipulating the length and data didn't really seem any cleaner. If you think this is a real use case that shouldn't be solved in some other way, I'll create a helper function :+1:

joerchan replied:
Double underscode (__buf) means that you are dealing with internal variables and that you should really know what you are doing. So I would think that in this case maybe the best thing would be to create a function in net/buf for what you want to do. Because you should not have this dependency into your part of the code.

jhedberg commented on subsys/bluetooth/mesh/friend.c:
```diff
@@ -316,13 +322,124 @@ static struct net_buf *create_friend_pdu(struct bt_mesh_friend *frnd,
+	}
+
+	err = unseg_app_sdu_unpack(frnd, buf, &meta);
+
```
Remove this empty line. Basically, to be consistent with style, whenever assigning to a variable and then testing for the value there should be no empty line in between.

jhedberg commented on subsys/bluetooth/mesh/friend.c:
```diff
@@ -316,13 +322,124 @@ static struct net_buf *create_friend_pdu(struct bt_mesh_friend *frnd,
+	}
+
+	err = unseg_app_sdu_decrypt(frnd, buf, &meta);
+
```
Same here

jhedberg commented on subsys/bluetooth/mesh/friend.c:
```diff
@@ -316,13 +322,124 @@ static struct net_buf *create_friend_pdu(struct bt_mesh_friend *frnd,
+	}
+
+	err = unseg_app_sdu_encrypt(frnd, buf, &meta);
+	if (err) {
```
Here you got it right. See, you're even inconsistent with yourself 😜

jhedberg commented on subsys/bluetooth/mesh/friend.c:
```diff
@@ -1217,6 +1346,15 @@ static void friend_lpn_enqueue_tx(struct bt_mesh_friend *frnd,
+		FRIEND_ADV(buf)->app_idx = tx->ctx->app_idx;
+	}
+
+
```
One redundant empty line here (there should only ever be one consecutively)

jhedberg commented on subsys/bluetooth/mesh/friend.c:
```diff
@@ -1278,6 +1278,10 @@ static void friend_lpn_enqueue_rx(struct bt_mesh_friend *frnd,
 	struct friend_pdu_info info;
 	struct net_buf *buf;
 
+	if (bt_mesh_elem_find(rx->ctx.addr)) {
```
Maybe a code comment here? Essentially the same as you have in the commit message. Is this a bug fix btw, or something that became necessary with this PR?

trond-snekvik replied:
This is a bug fix that is necessary for this patch to work correctly.

jhedberg commented on subsys/bluetooth/mesh/friend.c:
```diff
@@ -316,13 +322,124 @@ static struct net_buf *create_friend_pdu(struct bt_mesh_friend *frnd,
+	int err;
+
+	meta->subnet = bt_mesh_subnet_get(frnd->net_idx);
+	meta->is_dev_key = (FRIEND_ADV(buf)->app_idx == BT_MESH_KEY_DEV);
```
You could use here the local `app_idx` variable (since you have it and it'll produce a shorter line)

trond-snekvik committed:
net: buf: Add net_buf_simple_clone

Provides a way to clone a net_buf_simple without altering the state of
the original buffer. The primary usage scenario is for manipulating a
previously allocated PDU inside a buffer without altering the length and
offset of the buffer.

Signed-off-by: Trond Einar Snekvik <Trond.Einar.Snekvik@nordicsemi.no>

trond-snekvik committed:
Bluetooth: Mesh: Use net_buf_simple_clone

Uses net_buf_simple_clone to access the sdu of an unsegmented app packet
for re-encryption.

Signed-off-by: Trond Einar Snekvik <Trond.Einar.Snekvik@nordicsemi.no>

jhedberg commented on include/net/buf.h:
```diff
@@ -153,6 +153,27 @@ static inline void net_buf_simple_reset(struct net_buf_simple *buf)
+ * @param offset Offset relative to the current buffer offset.
+ * @param len Length of the sub buffer.
+ */
+static inline void net_buf_simple_scope(struct net_buf_simple *scope,
```
We have `net_buf_clone()` which copies the state of another buffer but only adds a reference to the payload if the buffer type supports that. I think we could call this new API `net_buf_simple_clone()` but then the result would need to reflect the parsing state of the original. I.e. after calling it you'd need to still adjust the length & offset.

jukkar replied:
`net_buf_simple_clone()` sounds find to me

trond-snekvik replied:
Changed it to _clone() that just shallow copies the buffer state, moving the offset and length of the sdu afterwards.

jhedberg commented on include/net/buf.h:
```diff
@@ -153,6 +153,20 @@ static inline void net_buf_simple_reset(struct net_buf_simple *buf)
 	buf->data = buf->__buf;
 }
 
+
```
Unnecessary empty line here

jhedberg commented on include/net/buf.h:
```diff
@@ -153,6 +153,20 @@ static inline void net_buf_simple_reset(struct net_buf_simple *buf)
+ * @param original Buffer to clone.
+ */
+void net_buf_simple_clone(struct net_buf_simple *clone,
+			  const struct net_buf_simple *original);
```
I'd swap the parameter order since `original` is "the" object you're cloning and `clone` is the output/result of the function. Thoughts? (I'm not completely convinced either way)

trond-snekvik replied:
Either way is fine by me. I used this order to match memcpy, as that's the stdlib equivalent of this, but either is fine by me.

jhedberg replied:
Let's swap. The primary object that the APIs operate on is usually the first parameter. Or to put it another way, for an API called `net_buf_simple_<verb>()` the first parameter should be the answer to the question "what do you want to \<verb\>?".

jhedberg commented on subsys/bluetooth/mesh/friend.c:
```diff
@@ -1239,6 +1380,10 @@ static bool friend_lpn_matches(struct bt_mesh_friend *frnd, u16_t net_idx,
 
+	if (addr == BT_MESH_ADDR_ALL_NODES) {
+		return true;
+	}
```
Side note: I think we have the same bug in `client_filter_match()` in `mesh/proxy.c`

jhedberg reviewed (approved):
Looks good now. I'd perhaps still refrain from merging until we have full confidence in the all-nodes address handling. Alternatively feel free to split that out into its own PR so we could merge everything else.

trond-snekvik commented:
I'll pull out the all-nodes commit and make a separate PR if we decide we want to do it.
